Attn implementation for composite models

zucchini-nlp opened the pull request:
# What does this PR do?

Related to #32221 and fixes #30565. As described in the issue, currently MultiModal models set attn implementation only in the LM backbone, and not in vision backbone. While working on it, I found another bug. Specifically, using a property to set SDPA flag to VLM class doesn't work because we don't know what is `self.LM` before init the VLM. I tried class property, but that would throw error that `self has no attribute LM`. 

This PR makes a few modifications to how attn impl is set in modeling to fix the above issues.

More precisely:

- If the model is composite, i.e. has 2 or more PreTrainedConfigs as part of the main config (text/vision config), then we set the requested attention implementation to each of the sub-configs as a dict. For ex, in LLaVa case:
```python
model.from_pretrained(id, attn_implementation="sdpa")

# internally we do the following, where the second line dispatches SDPA on LM backbone
# if we can't dispatch, an error is raised as it should be when loading an LLM with unsupported attention (i.e. llama doesn't support SDPA)
self.config._attn_implementation = {"text_config": "sdpa", "vision_config": "sdpa"}
AutoModelForCausalLM._from_config(config.text_config, attn_implementation=config._attn_implementation["text_config"])
```

- If user just loads the composite model without any explicit attention, then we dispatch SDPA in any of the sub-models whenever possible. In case any/none of the vision/text models support SDPA, we silently fallback to eager. No warnings, no errors, which is in line with non-composite models
```python
# suppose LM backbone doesn't support SDPA here, then we'll use eager silently
# but dispatch SDPA on vision backbone
model.from_pretrained(id, attn_implementation=None)
```

- The users are also free now to set different attentions in different backbones. We can set eager on vision model and sdpa on LM. That can be done if one passes a dict `attn_implementation`. And the dict keys have to be identical to the ModelConfig attribute names for sub-configs. In other words, if the config has config.text_config, then the dict key must be "text_config" to correctly dispatch

```python
model.from_pretrained(id,  {"text_config": "sdpa", "vision_config": "eager"})
```
- For non-composite models nothing changes and everything works as before


NOTE:
I had a few rare cases where the model was not really composite but had several sub-configs. For example DBRX, and I decided to get rid of the sub-configs there. I am not sure what is the right way to deprecate that, so any comments on that welcome
Also, some models have nested configs, where a vision config is part of text config (Qwen2-VL). Only three models are like that, mostly because I didn;t review properly when the model was added and didn't insist on changing stuff. For those models, we don't consider them as composite and just leave it working as it was

zucchini-nlp committed:
first try

zucchini-nlp committed:
codestyle

zucchini-nlp committed:
idefics2 is happy

zucchini-nlp committed:
[run-slow] llava, llava_next, video_llava, vipllava, llava_next_video, idefics, idefics2, kosmos2, fuyu, blip, blip_2, instructblip, instructblipvideo, paligemma

zucchini-nlp committed:
fix-copies

zucchini-nlp committed:
[run-slow] llava, llava_next, video_llava, vipllava, llava_next_video, idefics, idefics2, kosmos2, fuyu, blip, blip_2, instructblip, instructblipvideo

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32238). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
blip-2 needs to init vision from config

qubvel reviewed (commented):
Thanks for working on this! Not sure I fully understand the logic at this moment, so I left a few comments below

zucchini-nlp committed:
when was this removed O_o

zucchini-nlp committed:
minor fix

zucchini-nlp commented on src/transformers/modeling_utils.py:
```diff
@@ -1512,6 +1513,40 @@ def _autoset_attn_implementation(
+            text_model_cls = MODEL_MAPPING.get(type(config.text_config), None)
+            if text_model_cls is not None:
+                config.text_config._attn_implementation = config._attn_implementation
+                config.vision_config._attn_implementation = config._attn_implementation
```
Right! Fixed that one! Would be nice to have a test on VLM side when vision-sdpa is not supported, let me see if I can arrange it

zucchini-nlp commented on src/transformers/modeling_utils.py:
```diff
@@ -1512,6 +1513,40 @@ def _autoset_attn_implementation(
+            if text_model_cls is not None:
+                config.text_config._attn_implementation = config._attn_implementation
+                config.vision_config._attn_implementation = config._attn_implementation
+                cls._supports_sdpa = text_model_cls._supports_sdpa
```
Yes, actually now I think we can get rid of `if text_cls` because text model is always there for sure. But vision can be a custom model which is not a `PretrainedModel` like in Idefics. The main idea is that the general VLM's `support_sdpa` flag should be set to `True` if at least one of the sub-configs have its `support_sdpa=True`

zucchini-nlp commented on src/transformers/modeling_utils.py:
```diff
@@ -1512,6 +1513,40 @@ def _autoset_attn_implementation(
             # If a config is passed with a preset attn_implementation, we skip the automatic dispatch and use the user-provided config, with hard checks that the requested attention implementation is available.
             requested_attn_implementation = config._attn_implementation_internal
 
+        # MultiModal LLM related hack: since they consist of two or might be more sub-models
+        # we have to check and dispatch SDPA to each sub-model, in case any of them support it.
+        # If one sub-model supports SDPA while other doesn't, an error will be raised following the
+        # typical SDPA-dispatch path.
+        # Same goes for the `_supports_cache_class` because we cannot know what flag LM has
+        # before knowing which class is that LM
```
Yes, so that if one sub-model can dispatch on SDPA, it will dispatch. But the other will not. At the same time the general `config._attn_implementation` flag in for ex. LLaVa will be "sdpa" , and will not be used while init

zucchini-nlp committed:
tests

qubvel commented on src/transformers/modeling_utils.py:
```diff
@@ -1512,6 +1513,40 @@ def _autoset_attn_implementation(
+            if vision_model_cls is not None:
+                config.vision_config._attn_implementation = config._attn_implementation
+                cls._supports_sdpa = vision_model_cls._supports_sdpa
+                cls._autoset_attn_implementation(
```
Should it be this way?
```diff
-                cls._autoset_attn_implementation(
+                vision_model_cls._autoset_attn_implementation(
```

zucchini-nlp replied:
Yes, this is actually better than the current way cause we don;t have to reset `supports_sdpa` flag every time

ydshieh commented on src/transformers/modeling_utils.py:
```diff
@@ -1512,6 +1513,40 @@ def _autoset_attn_implementation(
+                )
+
+            if vision_model_cls is not None and text_model_cls is not None:
+                cls._supports_sdpa = vision_model_cls._supports_sdpa or text_model_cls._supports_sdpa
```
It's very bad to modify a class attribute: it will lead bugs and nightmares 😰   very difficult to detect.

One example is the `InstructBlip`, see #27867 and #25910.

I see you need it in the call below to `_check_and_enable_sdpa`. Got to think another way ...

ydshieh replied:
BTW, 

```
        # If one sub-model supports SDPA while other doesn't, an error will be raised following the
        # typical SDPA-dispatch path.
```

If I understand correctly, the call to `_check_and_enable_sdpa`  and `_autoset_attn_implementation` will return without error, right? The error occur only when we try to init. the model with this config?

zucchini-nlp replied:
Yes, the error for SDPA is only when the user specifically sets `attn_implementation=sdpa` when loading the model. Dispatching SDPA on sub-configs of VLM will follow the same path so that we raise error that "VisionMode doesn't support sdpa" only if user specifically requested sdpa. Otherwise we dispath to eager by default

ydshieh commented:
Regarding

```
MODEL_MAPPING.get(type(config.text_config), None)
            if text_model_cls is not None:
```
it could probably reuse

```
def _get_model_class(config, model_mapping):
```

But both will be somehow slow as you mentioned, and sometimes give unexpected errors (because loading a modeling module file). But we might have someway to avoid that.

zucchini-nlp committed:
this way?

zucchini-nlp committed:
tests

zucchini-nlp commented:
@ydshieh @qubvel I made some changes after our last discussion that changing class attr is not good. So now we don't change the attr per se, but we change the config's `attn_impl` manually while setting sdpa. I checked that it works this way also. The last question is whether to set it to SDPA when both sub-configs support it or at least one supports it. 


And also, regarding
> But both will be somehow slow as you mentioned, and sometimes give unexpected errors (because loading a modeling module file). But we might have someway to avoid that.

unfortunately we can't do that, cause it raises error if key is not in the dict and we have not auto-mappable models. So we rather try to get and if not simple continue.

ydshieh commented:
> unfortunately we can't do that, cause it raises error if key is not in the dict and we have not auto-mappable models. So we rather try to get and if not simple continue.

but why not try except around the usage of `_get_model_class`? I don't feel very strong here, but just it is something well designed in order to get the model class which is good if we can reuse it.

zucchini-nlp commented:
Hmm, not sure how good if to use a `try except`, can it be another source of hidden bugs even if we use `except AttributeError`? I'm okay with any way, as long as it's consistent with the library and doesn't cause bugs

zucchini-nlp committed:
model-agnostic code

zucchini-nlp committed:
codestyle

zucchini-nlp committed:
add tests for idefics

zucchini-nlp committed:
modify general test for VLMs

zucchini-nlp committed:
no generation test for vlm yet!

zucchini-nlp commented:
As we discussed internally, I made the loading logic model-agnostic. So now we check if the config contains any `PretrainedConfig` inside and set attn implementation to each `PretrainedConfig`. This solved vision-encoder-decoder problem we had in slack and makes us more versatile for possible future multimodal models.

I added some tests in vision enc-dec models where only one sub-config supports SDPA and when both support. Also added tests in Idefics (which were skipped earlier) to verify that LM has its SDPA dispatched even though other modules do not.

TODO in the next few hours or tomorrow:
- Add similar tests to encoder-decoder models, not only vision enc-dec
- Accept from users a config with separate `attn_implementation` already set, such that `attn_implementation={"vision_config": "sdpa", "text_config": "eager"}`

zucchini-nlp committed:
no generation test here also

zucchini-nlp committed:
wanr in VIT-SDPA if output attn

zucchini-nlp committed:
add more tests

zucchini-nlp committed:
user can pass dict as attn impl

zucchini-nlp committed:
repo consistency

zucchini-nlp committed:
update

zucchini-nlp committed:
muicgen

zucchini-nlp committed:
no prints

zucchini-nlp committed:
forgot speech enc-dec and clip

zucchini-nlp committed:
how many composite models we have?

zucchini-nlp committed:
musicgen meelody is same as mudicgen

zucchini-nlp committed:
+siglip

zucchini-nlp committed:
fix tests + add some more

zucchini-nlp committed:
remove idefics custom overriden code

zucchini-nlp commented:
I made final changes to the PR and made sure all tests are passing and all composite models were modified. The main points are:

- we have a `Is-composite` flag for models so that we know if we want to dispatch on each sub-config or not. The flag was needed for models like Chameleon that have a sub-config, but it's for VAE which will not supports SDPA i guess
- attn implementation can be set as prev ("sdpa") or as a dict in case of composite models ({"encoder": "sdpa", "decoder": "eager"}). For composite models, in first case we try to dispatch requested attn in all sub-models and raise error if not supported. In second case we dispatch the different attn for each sub-model as requested and also raise error if not supported
- Idefics and MusicGen models can get rid of custom code for attn implementation now
- For testing I had to add another flag `is_multimodal` (maybe will change to `is_composite`) and check each sub-config's attn implementation is set correctly. Also added tests in enc-dec models
- For some models which were not automappable, I had to add them to mapping so that we can get from config to class, and obtain `supports_sdpa` flag


One last thing to consider is that this might result in a bit of slow-down as we try to do automapping for each sub-model and dispatch separately. I have another idea, we can stop dispatching in each sub-model and instead simply replace "str" attn implementation with "dict" as `attn_implementation = {"encoder": requested_attn, "decoder": requested_attn}`. Then inside modeling we could do `AutoModel.from_config(config, attn_impl = config._attn_impl["encoder"])`. I testes with a few models and verified it works same way as current PR but did not introduce those changes yet. First, I want to get a general feedback if the PR is doing what we want. cc @amyeroberts, would like to know if I addressed your concerns about hidden magic code

zucchini-nlp committed:
make idefics2 automappable

zucchini-nlp committed:
nits

zucchini-nlp committed:
skip tests

zucchini-nlp committed:
doctests

amyeroberts commented on tests/models/llava_next/test_modeling_llava_next.py:
```diff
@@ -208,14 +207,20 @@ def create_and_check_llava_next_model_fp16_autocast_forward(
 
 @require_torch
-class LlavaNextForConditionalGenerationModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
+class LlavaNextForConditionalGenerationModelTest(ModelTesterMixin, unittest.TestCase):
```
Why remove `GenerationTesterMixin`? I'm guessing because before `all_generative_model_classes` wasn't defined and this is multimodal but just want to make sure I've understood the change properly

amyeroberts reviewed (commented):
Really great piece of work - thanks for taking the time and energy to tackle this tricky behaviour and find such a nice solution! 

Some general comments about idefics and a iterating over the config. Otherwise looks great!

zucchini-nlp committed:
Update src/transformers/models/idefics2/configuration_idefics2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp commented on src/transformers/__init__.py:
```diff
@@ -6871,8 +6873,10 @@
         from .models.idefics2 import (
             Idefics2ForConditionalGeneration,
             Idefics2Model,
+            Idefics2PerceiverResampler,
             Idefics2PreTrainedModel,
             Idefics2Processor,
+            Idefics2VisionTransformer,
```
Yes, I wouldn't import it also. I think after the next iteration of changes I'll try to get rid of extra `AutoMap` for sub-configs, this can be removed. That works for most models, but haven't really thought of Idefics

zucchini-nlp committed:
Update tests/models/clip/test_modeling_clip.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update tests/models/idefics2/test_modeling_idefics2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update tests/models/idefics2/test_modeling_idefics2.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update src/transformers/configuration_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
major update, no need for automap

zucchini-nlp committed:
clean up

zucchini-nlp committed:
add FA2 test

zucchini-nlp committed:
more tests

zucchini-nlp committed:
merge main

zucchini-nlp committed:
style

zucchini-nlp commented:
Made updates and removed automapping, seems we don't need it to dispatch in submodules. Added more tests, specifically for FA2 and for the new way of setting attn impl (as a dict in composite models). All the tests I added/touched are passing for me locally + non-slow tests in all models.

Some FA2/SDPA tests are failing but those seem to be failing disregarding the current changes. Requesting re-review, failing tests are not related and should have been skipped

zucchini-nlp committed:
skip tests

zucchini-nlp committed:
why did these started failing now?

zucchini-nlp committed:
no attributes for FA2 needed

zucchini-nlp committed:
one tiny test

nbroad1881 commented:
Will [this warning](https://huggingface.co/Qwen/Qwen2-VL-7B-Instruct/discussions/23) stop appearing after this PR is merged?

zucchini-nlp commented:
@ncoop57 not really in the current state, because we always call `from_config(attn_implementation="my_attn_impl")` to initialize a sub-component and we don't indicate `dtype`. Actually, now that I look into that, seems like we could call simply `SubModule(config)` instead of `_from_config`. The only thing is that directly calling class won't work with deepspeed, according to code. I'll check this out, should be also fixed in scope of this PR

zucchini-nlp committed:
address comment about FA2 false warning

zucchini-nlp commented:
TODO: New VLMs were added while PR was in progress, so we need to propagate new design to:
- [ ] Qwen2-VL
- [ ] LLava-Onevision
- [ ] Idefics2
- [ ] Some more coming soon


In the meanwhile, @amyeroberts or @ArthurZucker , can I get a review on general design ?

amyeroberts reviewed (commented):
Overall structure looks great to me. 

Just some small comments on the PR atm, although I was scanning more high-level. 

One question I have is about the checks we have for an individual model versus composite models. My understanding at the moment is that if I have a single model where I do: `Model.from_pretrained(checkpoint, attn_implementation="flash_attention_2)"`, then we end up calling `Model._check_and_enable_flash_attn_2` which will verify if FA2 is available and if the model is FA2 compatible. 

Am I right in saying if `Model` is a composite model, e.g. with a vision block and a language block, and I called `Model.from_pretrained(checkpoint, attn_implementation="flash_attention_2)"` then: 
* Both blocks would have their attention assigned as flash attention 2?
* Verification would be run for those individual blocks respectively, as `cls._autoset_attn_implementation` is called in `_from_config`? i.e. if it's compatible in one block but not the other an error will still be raised?

zucchini-nlp commented:
>  i.e. if it's compatible in one block but not the other an error will still be raised?

Yes this is the behavior now. Sorry I forgot to update PR description so let me describe the expected behavior.

For non-composite models nothing changes and the attention is dispatched same way as before. For composite models:

- if user just loads the model without any explicit attention (`Model.from_pretrained(model_id)`), then we dispatch SDPA in any of the sub-models whenever possible. In case any/none of the vision/text models support SDPA, we silently fallback to eager. No warnings, no errors, which is in line with non-composite models
- if user specifies required attention, we check if all of the sub-models support it. An error will be raised on sub-model level if the attention is not yet supported, so instead of `Llava doesn't support FA2`, we'll raise `CLIP doesn't support FA2`
- if user specifies attention for each sub-model when loading, as a dict (`Model.from_pretrained(model_id, attn_implementation={"vision_config": "eager", "text_config": "sdpa"})`), then we dispatch required attention to indicated models. We raise an error if any of the requested attentions is not supported by the indicated sub-model. When using attn implementation as a dict the user has to indicate attn for all sub-models, no defaults will be available. And the dict keys have to be identical to the `ModelConfig` attributes for sub-configs. In other words, if the config has `config.text_config`, then the dict key must be `text_config` to correctly dispatch

zucchini-nlp commented on src/transformers/modeling_utils.py:
```diff
@@ -1441,7 +1444,7 @@ def _from_config(cls, config, **kwargs):
                 Override the default `torch.dtype` and load the model under this dtype.
         """
-        torch_dtype = kwargs.pop("torch_dtype", None)
+        torch_dtype = kwargs.pop("torch_dtype", torch.get_default_dtype())
```
because of the above comment about FA2 warnings that model dtype should be half-precision. When we call `_from_config` within modeling code, we don't pass `dtype`. In general it doesn't break anything and the correct dtype is used, but dispatching attention implementation checks for this particular kwarg from inputs. So when the kwarg is None, it assumed we are not in half-precision and raises a warning that "FA2 behavior might be bad and unexpected"

ArthurZucker replied:
let's add a small comment at the top!

zucchini-nlp replied:
that might go in a different PR from contributor, the one you linked to this PR

zucchini-nlp commented on src/transformers/configuration_utils.py:
```diff
@@ -831,6 +831,10 @@ def __eq__(self, other):
         return f"{self.__class__.__name__} {self.to_json_string()}"
 
+    def __iter__(self):
+        for attr in copy.deepcopy(self.__dict__):
+            yield attr
```
Hmm, it has been long since I made this. Let me check out if the requirement to cope still holds or we can get rid of it

zucchini-nlp commented on src/transformers/modeling_utils.py:
```diff
@@ -1525,6 +1532,28 @@ def _autoset_attn_implementation(
+                    else requested_attn_implementation.get(key)
+                )
+
+            if cls._is_composite:  # some composite models don't use attn impl, e.g. VQ-VAE
```
This is for specific cases when a model is composite (has several standalone models) but not all of the components are a Transformer. So the sub-component doesn't/can/t have any attention mechanism and has nothing to dispatch. 

Example: Chameleon with its VQ-VAE or CLIPText models. 

In case of CLIP, the model is mostly loaded with the general config (not text config). So when a `ClipTextModel.from_pretrained()` is loaded it relies on general `ClipConfig` and we replace the config's attention implementation with a dict for each sub-model (text vs vision config). In later steps we fail to dispatch because we are dispatching on `self.config._attn_implementation` and  `attention implementation should be a string and one of the following....`. Hope this explains why. I thinkg for Chameleon we could find a way without `is_composite` attribute but in case of CLIP-like models that would be harder

amyeroberts commented:
> * if user specifies required attention, we check if all of the sub-models support it. An error will be raised on sub-model level if the attention is not yet supported, so instead of `Llava doesn't support FA2`, we'll raise `CLIP doesn't support FA2`

@zucchini-nlp Great, thanks for explaining. Behaviour all sounds good to me!

zucchini-nlp committed:
merge main

zucchini-nlp committed:
style

zucchini-nlp committed:
add new models and resolve conflicts

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
let it be this way for now, come back tomorrow to review

zucchini-nlp commented on src/transformers/models/idefics2/modeling_idefics2.py:
```diff
@@ -923,10 +1000,15 @@ def __init__(self, config, layer_idx: int):
         self.n_latents = config.perceiver_config.resampler_n_latents
         self.depth = config.perceiver_config.resampler_depth
         self.rms_norm_eps = config.text_config.rms_norm_eps
+        attn_implementation = (
+            config._attn_implementation["perceiver_config"]
+            if config._attn_implementation["perceiver_config"] is not None
+            else "eager"
+        )
```
Couldn't make Perceiever a PreTrained model which mean we can't call "dispatch_attn" when loading the perceiver module. A bit hacky, but since we know perceiver supports all attn types we can just dispatch to whatever was asked by the user. No need for warnings/errors, because hardware related errors will be raise when dispatching the vision/;m backbones

The main reason why perceiver couldn't be a pretrained model, is because it accepts a general config, not PerceiverConfig as input. Later it uses perceiver config in attention modules

zucchini-nlp committed:
some more fixes

zucchini-nlp committed:
update

zucchini-nlp committed:
more updates

zucchini-nlp committed:
update

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm_sdpa_flag

zucchini-nlp committed:
style and tests

zucchini-nlp commented on src/transformers/models/idefics/modeling_idefics.py:
```diff
@@ -980,16 +980,46 @@ def _init_weights(self, module):
-
-        if not hard_check_only:
-            config._attn_implementation = "sdpa"
+    def _autoset_attn_implementation(
+        cls,
+        config,
+        use_flash_attention_2: bool = False,
```
overwrite to set "sdpa" no matter what because that is the only attention that IDEFICS has, and basically `attn_implementation` is not used anywhere except for when preparing the causal mask

zucchini-nlp commented:
I think now this is ready for review, @ArthurZucker 

I rebased main and updated all new models, addressed some comments from Amy and added a few more tests

ArthurZucker commented:
Related to #33953 as well, will have a look, but having to change modeling code means we are doing something wrong as it should be a layer of abstraction!

zucchini-nlp commented:
Cool thanks! The modeling is mostly changed to get the attn implementation from dict and for IDEFICS (1 and 2), which fall out of standards and enforce hard requirements on sdpa. Hope we can get this merged soon, as it is getting more of a problem when new quirky models are added

ArthurZucker commented on examples/modular-transformers/modeling_my_new_model2.py:
have seen these modified elsewhere, let's revert as unrelated

ArthurZucker commented on src/transformers/models/blip_2/modeling_blip_2.py:
```diff
@@ -1448,19 +1449,21 @@ class Blip2Model(Blip2PreTrainedModel):
         super().__init__(config)
 
-        self.vision_model = Blip2VisionModel(config.vision_config)
+        self.vision_model = Blip2VisionModel._from_config(
+            config.vision_config, attn_implementation=config._attn_implementation["vision_config"]
+        )
```
my main question is why is this info not stored in the config.vision_config directly? 
Initializing the composite config should propagate the `attn_implementation` attribute to the sub configs, instead of extracting it in modeling

ArthurZucker replied:
That means updating `from_config` to support extracting the attn_implementation from the config 😉

zucchini-nlp replied:
oh right! The was the initial plan was similar, but it was supposed to recursively call `autoset_attn_impl` from within `autoset_attn_impl` for each sub-module-class. But that was too much extra code and checks

But I think I got you mean here, like instead of doing a dict for generatl config we can simply propagare user-requested attn for each sub-config. Then we hope it will get picked up in later `AutoModel.from_config` calls. Good idea and a lot cleaner, let me see how it goes. Hopefully that way we don't need specific treatment for nested configs

ArthurZucker replied:
🤗 good luck!

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -1580,6 +1584,32 @@ def _autoset_attn_implementation(
             # If a config is passed with a preset attn_implementation, we skip the automatic dispatch and use the user-provided config, with hard checks that the requested attention implementation is available.
             requested_attn_implementation = config._attn_implementation_internal
 
+        # Composite models consisting of several PretrainedModels have to specify attention impl as a dict
+        # where keys are sub-config names. But most people will specify one `str` which means that should dispatch it
+        # for all sub-models.
```
It makes sense, sometimes you want output hidden states of one module but not the other one, so I guess it's minimul flexibility.

ArthurZucker replied:
but IMO we should propagate here

ArthurZucker reviewed (commented):
Thanks, it is indeed important work!

zucchini-nlp committed:
another big update

zucchini-nlp committed:
fix tests

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into vlm_sdpa_flag

zucchini-nlp committed:
fix tests

zucchini-nlp committed:
update

zucchini-nlp committed:
another update

zucchini-nlp committed:
Merge branch 'main' into vlm_sdpa_flag

zucchini-nlp committed:
fix tests

zucchini-nlp commented:
@ArthurZucker I made changes as you suggested but I want to warn that it still needed a few extra logic. Making this PR revealed soo many issues. So we have two options now to indicate attn for composite models:

**Attn implementation as a dict:**
The prev solution when we simply copy paste the user-requested attn in all sub-configs, and form a dict. So that the general `config.attn = Dict(text_config: None, vision_config: None)`
+ no need for flag whether attn implementation is set or not, because when we have a dict we always use the very first requested attn in all sub-configs. We just need to save it as a dict once and then go back to that dict to "get()"
+ we don't need to unset attn implementation when saving so one can set a different attn when loading any model back.
- BUT we change a lot of model code to pass in explictly (attn_implementation=config._attn_implementation["text_config"]) and we need to account for cases like Chameleon or DBRX where the sub-configs don't even need attn

**Attn implementation directly propagated in corresponding configs:**
Current solution when we simply propagate the attn to all configs we can find, and don't touch the general config
+ no need to change any modeling code, except for general clean up
- BUT we need a flag attn_implementation_autoset  so we don't dispatch attn on the same config twice. Also we would need to unset the flag when saving. See below for reason

Imagine we have a composite or nested config. Currently we set attn in three places: `from_pretrained`, `from_config` and `init`. When we don't require any attn (`None`) and load with `model.from_pretrained(id)`, internally we will call `autoset_attn`  and set SDPA to the general config and the sub-configs will get `None` propagated. But then the same method `autoset_attn`  is called second time when we init the class, and this time we'll try to propagate SDPA to all sub-configs (because general config has SDPA already set). If any sub-module has no SDPA support, an error will be raised as it was requested directly. From user perspective they did nothing wrong, but they see an error.
So the decision is to either stop autosetting attn for each config once it was set, which is happening now. Or in prev version I had attn implementations saved as dict, which basically did the same. When we got a dict attn, we had nothing to set and simply returned

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -1580,6 +1589,22 @@ def _autoset_attn_implementation(
+        # Below we check if a config is composite and manually prepare a dict of attn impl if not already passed as a dict.
+        # Later each sub-module will dispatch with its own attn impl, by calling `XXXModel._from_config(config.text_config)`
+        # If any of sub-modules doesn't support requested attn, an error will be raised. See https://github.com/huggingface/transformers/pull/32238
+        for key in config:
+            if isinstance(getattr(config, key), PretrainedConfig):
+                sub_config = getattr(config, key)
+                curr_attn_implementation = (
+                    requested_attn_implementation
+                    if not isinstance(requested_attn_implementation, dict)
+                    else requested_attn_implementation.get(key, None)
+                )
+                sub_config._attn_implementation_internal = curr_attn_implementation
```
If this is donc to set the `attn_implementation` of the sub configs, I am wondering if we can't have a way to use `is_composition` to simply say: 
- if we are a composition, then we call on our composiite member the `_autoset_attn_implementation` function

Today we have a problem with `is_compoisition`: we have this 
```python
    @classmethod
    def from_pretrained(cls, pretrained_model_name_or_path: Union[str, os.PathLike], **kwargs) -> "PretrainedConfig":
        cls._set_token_in_kwargs(kwargs)

        config_dict, kwargs = cls.get_config_dict(pretrained_model_name_or_path, **kwargs)

        if config_dict.get("model_type") == "mllama":
            config_dict = config_dict["vision_config"]

        if "model_type" in config_dict and hasattr(cls, "model_type") and config_dict["model_type"] != cls.model_type:
            logger.warning(
                f"You are using a model of type {config_dict['model_type']} to instantiate a model of type "
                f"{cls.model_type}. This is not supported for all configurations of models and can yield errors."
            )

        return cls.from_dict(config_dict, **kwargs)
```
which IMO we should never have to do and we copy past this everywhere.

This means there is something wrong with how we handle composition!

ArthurZucker reviewed (approved):
Modeling changes are just perfect 😉 before we have a big refactor this is super super welcome

zucchini-nlp committed:
merge main

zucchini-nlp committed:
fix copies

zucchini-nlp committed:
fix tests

zucchini-nlp commented:
merging, the tests pass locally in most cases except fro the cases not touched by this PR. For ex some FA2/sdpa equivalence tests are failing for me in `main` branch
